### PR TITLE
Added support for project users and roles

### PR DESCRIPTION
**Pull request by jakub-plichta**

### CODE REVIEW

File: src/main/java/com/gooddata/project/ProjectService.java [hunk 1]
```diff
@@ -254,4 +258,78 @@ public void handlePollException(final GoodDataRestException e) {
...
+        } catch (GoodDataException | RestClientException e) {
+            throw new GoodDataException("Unable to list users", e);
+        }
+    }
+
+    private static URI getUsersUri(Project project) {
+        return Users.TEMPLATE.expand(project.getId());
+    }
+
+    public Set<Role> getRoles(final Project project) {
```
Comment (martiner) [inline src/main/java/com/gooddata/project/ProjectService.java]: java doc

File: src/main/java/com/gooddata/project/ProjectService.java [hunk 2]
```diff
@@ -254,4 +258,78 @@ public void handlePollException(final GoodDataRestException e) {
...
+        notNull(project, "project");
+        final Roles roles = restTemplate.getForObject(Roles.URI, Roles.class, project.getId());
+        final HashSet<Role> result = new HashSet<>();
```
Comment (martiner) [inline src/main/java/com/gooddata/project/ProjectService.java]: `Set` interface

File: src/main/java/com/gooddata/project/Role.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    public String getTitle() {
+        return meta.getTitle();
+    }
+
+    public String getIdentifier() {
+        return meta.getIdentifier();
+    }
+
+    void setUri(final String uri) {
```
Comment (martiner) [inline src/main/java/com/gooddata/project/Role.java]: add a comment this is intentional

File: src/main/java/com/gooddata/project/Role.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    @JsonCreator
+    Role(@JsonProperty("permissions") final Map<String, Boolean> permissions,
+         @JsonProperty("meta") final Meta meta,
+         @JsonProperty("links") final Map<String, String> links) {
+        this.permissions = permissions == null ? new HashMap<String, Boolean>() : permissions;
+        this.meta = meta == null ? new Meta(null) : meta;
+        this.links = links == null ? new HashMap<String, String>() : links;
+    }
+
+    public Map<String, Boolean> getPermissions() {
```
Comment (martiner) [inline src/main/java/com/gooddata/project/Role.java]: i'm not sure this should be exposed to the poor user

File: src/main/java/com/gooddata/project/ProjectService.java [hunk 5]
```diff
@@ -254,4 +258,78 @@ public void handlePollException(final GoodDataRestException e) {
...
+        for (String roleUri : roles.getRoles()) {
+            final Role role = restTemplate.getForObject(roleUri, Role.class);
```
Comment (martiner) [inline src/main/java/com/gooddata/project/ProjectService.java]: this is n+1, is it necessary?

Comment (jakub-plichta) [inline src/main/java/com/gooddata/project/ProjectService.java]: I know, but there is no other way how to retrieve project roles

File: src/main/java/com/gooddata/project/Role.java (hunk 4 above)
Comment (jakub-plichta) [inline src/main/java/com/gooddata/project/Role.java]: well imagine you are developing some kind of administration UI - do you want to hardcode all currently supported permissions?

Comment (martiner) [inline src/main/java/com/gooddata/project/Role.java]: definitely not, but the `Map` is a bit weird. What about `Set<String> getEnabledPermissions()` and `Set<String> getAvailablePermissions()`?

Comment (jakub-plichta) [inline src/main/java/com/gooddata/project/Role.java]: :+1:  sounds good

File: src/main/java/com/gooddata/project/ProjectService.java (hunk 1 above)
Comment (jakub-plichta) [inline src/main/java/com/gooddata/project/ProjectService.java]: fixed

File: src/main/java/com/gooddata/project/ProjectService.java (hunk 2 above)
Comment (jakub-plichta) [inline src/main/java/com/gooddata/project/ProjectService.java]: fixed

File: src/main/java/com/gooddata/project/Role.java (hunk 3 above)
Comment (jakub-plichta) [inline src/main/java/com/gooddata/project/Role.java]: comment added

File: src/main/java/com/gooddata/project/Role.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    public boolean hasPermissionGranted(final String permission) {
+        return permissions.get(permission);
+    }
+
+    /**
+     * Grants provided permission.
+     *
+     * @param permission to be granted
+     */
+    public void grantPermission(final String permission) {
```
Comment (martiner) [inline src/main/java/com/gooddata/project/Role.java]: this is a bit misleading, there is no method for updating permissions, is it?

File: src/main/java/com/gooddata/project/Role.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        permissions.put(permission, true);
+    }
+
+    /**
+     * Revokes provided permission.
+     *
+     * @param permission to be revoked
+     */
+    public void revokePermission(final String permission) {
```
Comment (martiner) [inline src/main/java/com/gooddata/project/Role.java]: same as `grantPermission()`

**Comment by martiner** [comment]
Closes #143

File: src/main/java/com/gooddata/project/Role.java (hunk 6 above)
Comment (jakub-plichta) [inline src/main/java/com/gooddata/project/Role.java]: ah, that's right.. Thanks for noticing.

**Pull request merged**